api/src/main/java/io/druid/data/input/FiniteFirehoseFactory.java

```diff
@@ -0,0 +1,64 @@
+/*
+ * Licensed to Metamarkets Group Inc. (Metamarkets) under one
+ * or more contributor license agreements. See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership. Metamarkets licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License. You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied. See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package io.druid.data.input;
+
+import com.fasterxml.jackson.annotation.JsonIgnore;
+import io.druid.data.input.impl.InputRowParser;
+
+import java.io.IOException;
+import java.util.stream.Stream;
+
+/**
+ * {@link FirehoseFactory} designed for batch processing. Its implementations assume that the amount of inputs is
+ * limited.
+ *
+ * @param <T> parser type
+ * @param <S> input split type
+ */
+public interface FiniteFirehoseFactory<T extends InputRowParser, S> extends FirehoseFactory<T>
+{
+  /**
+   * Returns true if the {@link FirehoseFactory} supports parallel batch indexing.
+   */
+  @JsonIgnore
+  @Override
+  default boolean isSplittable()
+  {
+    return true;
+  }
+
+  /**
+   * Returns an iterator of {@link InputSplit}s.
+   */
+  @JsonIgnore
+  Stream<InputSplit<S>> getSplits() throws IOException;
```

Comment:
The mixing of streaming mechanics is confusing here. Why use `Stream` but not `Spliterator`?

Reply:
`SplitIterator` is designed to be tightly coupled with parallel processing itself. It has an interface accepting `Consumer`, so people can define an action to process each element in the collection and pass it to the `Spliterator`. This is not feasible in Druid, because the action to process each element is running tasks. 

Using `Stream` here is for some `FiniteFirehose`s which can incur high overhead to list all splits like `StaticS3FiniteFirese`. However, listing splits is not fully optimized in this PR because it should be done separately. I added this to the Javadoc of this method.

BTW, I realized some Javadoc comments in this class have gone stale. Updated them.

Reply:
Ok that makes sense